### PR TITLE
[clang] Fix a bug with qualified name lookup into current instantiation

### LINKED ISSUE #13826: Ignores dependent base classes using qualified lookup in definition context
|  |  |
| --- | --- |
| Bugzilla Link | [13454](https://llvm.org/bz13454) |
| Version | trunk |
| OS | Windows NT |
| CC | @DougGregor |

## Extended Description 
This should work fine, but Clang doesn't accept the absence of "typename" in the definition of "B"

template<typename T> struct A { 
   typedef int type; 
   struct B; 
}; 

template<typename T> struct A<T>::B : A<T> { 
   B::type t; 
};

### PR DESCRIPTION
**Pull request by Fznamznon**
Due to d0d2ee0e4bbe915d649e983c12d37bcfcf58823c clang doesn't perform qualified name lookup into the current instantiation when it has dependent bases, because of that `getTypeName` call always returns null for unknown specialization case. When there is a `typename` keyword, `DependentNameType` is constructed instead of simply returning null.
This change attempts to do the same in case of `typename` absence.

Fixes https://github.com/llvm/llvm-project/issues/13826

### CODE REVIEW

**Review by erichkeane — APPROVED** [review]
The patch makes sense to me, though my knowledge of this function is limited.  Please give @cor3ntin a chance to look at it too in case he sees something I don't.

File: clang/lib/Sema/SemaDecl.cpp [hunk 1]
```diff
@@ -482,8 +481,23 @@ ParsedType Sema::getTypeName(const IdentifierInfo &II, SourceLocation NameLoc,
...
+      T = Context.getDependentNameType(ElaboratedTypeKeyword::None,
+                                       SS->getScopeRep(), &II);
+      TypeLocBuilder TLB;
+      DependentNameTypeLoc TL = TLB.push<DependentNameTypeLoc>(T);
+      TL.setElaboratedKeywordLoc(SourceLocation());
+      TL.setQualifierLoc(SS->getWithLocInContext(Context));
+      TL.setNameLoc(NameLoc);
+      return CreateParsedType(T, TLB.getTypeSourceInfo(Context, T));
+    }
+    Result.suppressDiagnostics();
```
Comment (shafik) [inline clang/lib/Sema/SemaDecl.cpp:498]: We could also just do `[[fallthrough]];` at the end. I think that is fine.

File: clang/lib/Sema/SemaDecl.cpp [hunk 2]
```diff
@@ -482,8 +481,23 @@ ParsedType Sema::getTypeName(const IdentifierInfo &II, SourceLocation NameLoc,
...
       }
     }
-    // If typo correction failed or was not performed, fall through
-    [[fallthrough]];
+    Result.suppressDiagnostics();
+    return nullptr;
+  case LookupResult::NotFoundInCurrentInstantiation: {
+    if (AllowImplicitTypename == ImplicitTypenameContext::Yes) {
+      QualType T;
+      T = Context.getDependentNameType(ElaboratedTypeKeyword::None,
```
Comment (shafik) [inline clang/lib/Sema/SemaDecl.cpp:489]: Why not just initialize `T` when we declare it?

File: clang/lib/Sema/SemaDecl.cpp [hunk 3]
```diff
@@ -482,8 +481,23 @@ ParsedType Sema::getTypeName(const IdentifierInfo &II, SourceLocation NameLoc,
...
+                                       SS->getScopeRep(), &II);
+      TypeLocBuilder TLB;
+      DependentNameTypeLoc TL = TLB.push<DependentNameTypeLoc>(T);
+      TL.setElaboratedKeywordLoc(SourceLocation());
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaDecl.cpp:493]: I don't think that's actually needed even if we do it in a bunch of places.

**Review by cor3ntin — APPROVED** [review]
That looks reasonable to me modulo nit and @shafik comments.
Thanks for fixing this!

File: clang/lib/Sema/SemaDecl.cpp (hunk 3 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaDecl.cpp]: Oh, we probably need this. Sanitizer build failed due to elaborated keyword loc not initialized https://lab.llvm.org/buildbot/#/builders/5/builds/38796/steps/2/logs/stdio .

**Comment by Fznamznon** [comment]
Fix for buildbot https://github.com/llvm/llvm-project/pull/73928
I'm not able to verify locally though

**Comment by john-brawn-arm** [comment]
As a result of this commit clang no longer gives an error for
```
template <int A> struct X {
  enum E { a };
};

template <int A> struct Y : X<A> {
  Y::E m;
};
```
where gcc (though not msvc) gives an error: https://godbolt.org/z/qGfnzhfsK

**Comment by john-brawn-arm** [comment]
I'm not completely sure if we should be giving an error for the example I gave above. Looking through the C++20 standard I see temp.res paragraph 2:
>  A name used in a template declaration or definition and that is dependent on a template-parameter is assumed
not to name a type unless the applicable name lookup finds a type name or the name is qualified by the
keyword typename

The error that gcc is giving is
```
<source>:6:6: error: invalid use of incomplete type 'struct Y<A>'
    6 |   Y::E m;
      |      ^
<source>:5:25: note: definition of 'struct Y<A>' is not complete until the closing brace
    5 | template <int A> struct Y : X<A> {
      |                         ^
```
so I'm guessing that gcc thinks that the name lookup for ``Y::E`` doesn't find a type name because ``Y`` is incomplete, though I don't understand why the same isn't true for the example in 
```

**Comment by perry-ca** [comment]
Was this PR meant to change the behaviour for c++11/14?

**Comment by zygoloid** [comment]
> I'm not sure I understand why gcc is giving an error in https://godbolt.org/z/qGfnzhfsK. It doesn't give an error with `typename` keyword and I think there shouldn't be any difference since `Y::E` refers to current instantiation and is being used by class member declaration so no `typename` required. cc @shafik @cor3ntin @zygoloid for an opinion.

This looks like a GCC bug to me. I think the presence or absence of `typename` isn't supposed to make a difference in an implicit typename context.

**Pull request merged**